[KMSv2] feat: implements metrics for encryption config hot reload

nilekhc opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->
/kind feature

#### What this PR does / why we need it:

This PR implements metrics that record the total number of successful and failed encryption config hot reloads, along with the timestamp of the most recent successful hot reload.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #113704

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
New Metrics Added for Encryption Configuration Controller

This release adds new metrics to the Encryption Configuration Controller to help monitor the automatic reloading of encryption configuration. The new metrics include:

- `apiserver_encryption_config_controller_automatic_reload_failures_total`: Total number of failed automatic reloads of encryption configuration.
- `apiserver_encryption_config_controller_automatic_reload_success_total`: Total number of successful automatic reloads of encryption configuration.
- `apiserver_encryption_config_controller_automatic_reload_last_timestamp_seconds`: Timestamp of the last successful or failed automatic reload of encryption configuration.

These metrics can be used to monitor the health of the Encryption Configuration Controller and to troubleshoot any issues that may arise during automatic reloading of encryption configuration.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

/sig auth
/triage accepted
/cc @ritazh @enj @aramase

nilekhc commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/metrics/metrics_test.go:
I thought about writing table tests, but could not think of multiple test cases. I am open to suggestions.

ritazh commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/metrics/metrics.go:
```diff
@@ -0,0 +1,80 @@
+			Name:           "encryption_config_reload_last_success_timestamp_seconds",
+			Help:           "Timestamp of the last successful hot reload of encryption configuration.",
+			StabilityLevel: metrics.ALPHA,
+		},
```
How about if this Gauge metric records last time in seconds when a reload was performed and its status {success, fail}?

nilekhc replied:
We want to record the timestamp for the last successful hot reload. If the reload fails, then we don't care about when it failed. We just increment the failure count.  Is there any benefit to record a timestamp for failure as well?

ritazh replied:
If this metric is used to create alerts, then having the timestamp for the last failure is also helpful to know how fresh this failure is.

nilekhc replied:
makes sense. updated.

mengjiao-liu commented:
/sig instrumentation

aramase commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/controller/controller.go:
```diff
@@ -163,7 +164,7 @@ func (d *DynamicKMSEncryptionConfigContent) processNextWorkItem(serverCtx contex
 
 	defer func() {
-		// TODO: increment success metric when updatedEffectiveConfig=true
+		metrics.RecordEncryptionConfigHotReloadSuccess()
```
The `err != nil` check is done later, so this would update the success metric even if there is a failure?

nilekhc replied:
update the logic

aramase commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/metrics/metrics.go:
```diff
@@ -0,0 +1,80 @@
+
+const (
+	namespace = "apiserver"
+	subsystem = "encryptionconfig_controller"
```
```suggestion
	subsystem = "encryption_config_controller"
```

Consistent naming in the metrics

aramase commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/metrics/metrics.go:
```diff
@@ -0,0 +1,80 @@
+		&metrics.CounterOpts{
+			Namespace:      namespace,
+			Subsystem:      subsystem,
+			Name:           "encryption_config_hot_reload_failures_total",
```
```suggestion
			Name:           "hot_reload_failures_total",
```

This metric would be `apiserver_encryptionconfig_controller_encryption_config_hot_reload_failures_total` which has `encryption_config` twice.

aramase commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/metrics/metrics.go:
```diff
@@ -0,0 +1,80 @@
+		&metrics.CounterOpts{
+			Namespace:      namespace,
+			Subsystem:      subsystem,
+			Name:           "encryption_config_hot_reload_success_total",
```
```suggestion
			Name:           "hot_reload_success_total",
```

nilekhc committed:
feat: implements metrics for encryption config hot reload

Signed-off-by: Nilekh Chaudhari <1626598+nilekhc@users.noreply.github.com>

enj commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/metrics/metrics.go:
```diff
@@ -0,0 +1,86 @@
+		&metrics.CounterOpts{
+			Namespace:      namespace,
+			Subsystem:      subsystem,
+			Name:           "hot_reload_success_total",
```
```suggestion
			Name:           "automatic_reload_success_total",
```

enj commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/metrics/metrics.go:
```diff
@@ -0,0 +1,86 @@
+			Namespace:      namespace,
+			Subsystem:      subsystem,
+			Name:           "hot_reload_failures_total",
+			Help:           "Total number of failed hot reloads of encryption configuration.",
```
```suggestion
			Help:           "Total number of failed automatic reloads of encryption configuration.",
```

enj commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/metrics/metrics.go:
```diff
@@ -0,0 +1,86 @@
+			Namespace:      namespace,
+			Subsystem:      subsystem,
+			Name:           "hot_reload_success_total",
+			Help:           "Total number of successful hot reloads of encryption configuration.",
```
```suggestion
			Help:           "Total number of successful automatic reloads of encryption configuration.",
```

enj commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/metrics/metrics.go:
```diff
@@ -0,0 +1,86 @@
+		&metrics.GaugeOpts{
+			Namespace:      namespace,
+			Subsystem:      subsystem,
+			Name:           "hot_reload_latest_timestamp_seconds",
```
```suggestion
			Name:           "automatic_reload_last_timestamp_seconds",
```

enj commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/controller/controller.go:
```diff
@@ -163,16 +164,16 @@ func (d *DynamicKMSEncryptionConfigContent) processNextWorkItem(serverCtx contex
 		if !updatedEffectiveConfig {
 			// avoid leaking if we're not using the newly constructed transformers (due to an error or them not being changed)
 			closeTransformers()
+		} else {
+			metrics.RecordEncryptionConfigHotReloadSuccess()
```
This is easier to reason about:

```suggestion
		}
		if updatedEffectiveConfig && err == nil {
			metrics.RecordEncryptionConfigHotReloadSuccess()
```

enj commented:
This needs a release note that describes the newly added metrics.

nilekhc commented:
> This needs a release note that describes the newly added metrics.

Thanks for the review @enj. I have added the release note and addressed your comments. I'll squash commits if all looks good. PTAL.

logicalhan reviewed (commented):
/lgtm
/approve 

(from instrumentation)
/triage accepted

enj commented:
/lgtm
/approve

nilekhc commented:
/test pull-kubernetes-e2e-kind-kms

nilekhc commented:
Hi @dims, Could you PTAL for dependency change?  Thanks.

dims commented:
/lgtm
/approve
